README.md
# randomsound
[Randomsound](http://archive.ubuntu.com/ubuntu/pool/universe/r/randomsound/randomsound_0.2.orig.tar.gz) fills the entropy pool using data derived from a sound card.
This version of randomsound, which writes data from a sound card to a file, can be used to verify the randomness produced by the sound card.
Data is saved in a file in the randomsound directory.

## Introduction
Randomness in a computer is in general deterministic.
Mixing the randomness from a deterministic algorithm with non-deterministic data from some "device" increases the security of cryptographic algorithms and methods.
Examples for such "devices" are files, keyboard and mouse input, internet traffic or specific hardware devices like a sound card.
However, whether the device is correctly configured is not always clear, even though the software seems to work correctly.
This version of randomsound writes a file (rsound.out) in the randomsound directory which contains the data directly derived from the sound card for further analysis.
The file should be deleted after each run or the original randomsound program should be used if sound card was configured correctly.

Deterministic randomness is generated from /dev/urandom.
Non-deterministic randomness is kept in the entropy pool and provided by /dev/random.
See [this](https://stackoverflow.com/questions/23712581/differences-between-random-and-urandom) for the difference.
As a result:
(i) /dev/urandom always produces enough randomness but is vulnerable to cryptographic attacks;
(ii) /dev/random provides non-deterministic randomness but entropy pool might run empty, so that the device locks;

This implies, that a secure system should use /dev/urandom combined with a full entropy pool that is continously filled with high quality randomness.

## Methods
Nothing special.

## Examples:
### Start randomsound
Program must be run as root. Exit the program via Ctrl+C.

```
sudo ./randomsound -v
Random sound daemon. Copyright 2007 Daniel Silverstone.

Will keep random pool between 256 and 3840 bits of entropy.
Will retain a buffer of 1024 bytes, making entropy deposits of 64 bytes at a time.
Writing 512 bits of entropy to file
Writing 512 bits of entropy to file
Writing 512 bits of entropy to file
Did it fail?!?!
Writing 512 bits of entropy to file
Did it fail?!?!
Writing 512 bits of entropy to file
Writing 512 bits of entropy to file
...
Writing 512 bits of entropy to file
Writing 512 bits of entropy to file
Writing 512 bits of entropy to file
```
While the program is running, request random data (see next point).

### Request random data
In this case, use /dev/random and NOT /dev/urandom (see [this](https://stackoverflow.com/questions/23712581/differences-between-random-and-urandom) for the difference)!
Might take some time.

```
head -c 500 /dev/random > /dev/null
```

### Analyze derived data
Needs the program [ent](https://packages.ubuntu.com/de/xenial/ent).
Can be installed via apt-get (ubunutu).
The closer the resulting entropy is to 8 bits per byte the better. 
```
cat rsound.out | ent
Entropy = 7.907658 bits per byte.

Optimum compression would reduce the size
of this 3136 byte file by 1 percent.

Chi square distribution for 3136 samples is 411.76, and randomly
would exceed this value less than 0.01 percent of the times.

Arithmetic mean value of data bytes is 126.2790 (127.5 = random).
Monte Carlo value for Pi is 2.973180077 (error 5.36 percent).
Serial correlation coefficient is 0.074282 (totally uncorrelated = 0.0).
```


Changes
0.2.1	Writes data read from sound card to file for further analysis
0.2
	Incorporated manual page provided by Debian, updated it slightly.
	Added changes document.

0.1
	Numbered release, added -V option

Unnumbered release
	Initial release, fully functional


asoundrunner.c
#include "asoundrunner.h"

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/wait.h>
#include <signal.h>

#include <unistd.h>
#include <stdlib.h>
#include <fcntl.h>

#define SAMPLERATE 8000
#define SAMPLERATESTR "8000"

int arecord_read_fd = -1;
int restart_arecord = 1;

static BitBuffer read_into_here;
static int signal_handler_installed = 0;

void
child_handler(int _ignored)
{
  int status;
  while (waitpid(-1, &status, WNOHANG) != -1) ;
  restart_arecord = 1;
}

int
start_arecord(BitBuffer read_into)
{
  pid_t child;
  int pipefd[2];
  int sparefd;
  read_into_here = read_into;
  if (signal_handler_installed == 0) {
    signal(SIGCHLD, child_handler);
    signal_handler_installed = 1;
  }
  if (arecord_read_fd != -1) {
    close(arecord_read_fd);
    arecord_read_fd = -1;
  }
  if (socketpair(AF_LOCAL, SOCK_STREAM, PF_LOCAL, pipefd) == -1)
    return -1;
  restart_arecord = 0;
  child = fork();
  switch (child) {
  case -1:
    return -1;
  case 0:
    sparefd = open("/dev/null", O_RDWR);
    close(0);
    close(1);
    close(2);
    close(pipefd[0]);
    dup2(pipefd[1], 1);
    close(pipefd[1]);
    dup2(sparefd, 0);
    dup2(sparefd, 2);
    close(sparefd);
    execlp("arecord", "arecord", "-c", "1", "-f", "S16_LE", "-r", SAMPLERATESTR, "-t", "raw", NULL);
    exit(1);
  default:
    close(pipefd[1]);
    arecord_read_fd = pipefd[0];
  }
  return child;
}

void
asound_do_read(void)
{
  /* Read in SAMPLERATE*2/10 bytes (1/10th of a second of sampling) */
  BitField buffer[SAMPLERATE*5];
  int xfered = 0;
  int readbytes = read(arecord_read_fd, buffer, SAMPLERATE*5);
  if (readbytes < 0) {
    perror("read");
    restart_arecord = 1;
    return;
  }
  while (xfered < (readbytes>>1) && bitbuffer_free_space(read_into_here) > 0) {
    BitField thisbit = buffer[xfered];
    bitbuffer_add_bits(read_into_here, thisbit, 8);
    xfered += 2;
  }
}


randomsound.c
#include <sys/select.h>
#include <sys/times.h>
#include <sys/types.h>
#include <unistd.h>
#include <stdio.h>
#include <fcntl.h>

#include <linux/types.h>
#include <linux/random.h>
#include <sys/ioctl.h>

#include "bitbuffer.h"
#include "debias.h"
#include "asoundrunner.h"

static char *version_number = "0.2";
int daemonise = 0;
int verbose = 0;
int minwatermark = 256;
int maxwatermark = 4096-256;
int depositsize = 64;
int buffersize = 64 * 16;

BitBuffer incoming_bits;
BitBuffer buffered_bits;

int randomfd;

struct injector {
  int ent_count;
  int size;
  union {
    int ints[128];
    BitField bitfield[512];
  } value;
} random_injector;

int
bits_in_pool(void)
{
  int ret;
  ioctl(randomfd, RNDGETENTCNT, &ret);
  return ret;
}

void
do_mixin_bits(void)
{
  int i;
  int bits_before = bits_in_pool();
  if (verbose > 3)
    printf("Injecting %d bits of entropy into the kernel\n", depositsize * 8);
  /*
    with the correct soundcard setting, "depositsize * 8" should work correctly
    however, maybe it would be more secure to change "8" to a smaller value like 6 or 7, see also
    https://stackoverflow.com/questions/17118705/using-rndaddentropy-to-add-entropy-to-dev-random
  */
  random_injector.ent_count = depositsize * 8;
  random_injector.size = depositsize;
  for (i = 0; i < depositsize; ++i) {
    bitbuffer_extract_bits(buffered_bits, random_injector.value.bitfield + i, 8);
  }
  if (ioctl(randomfd, RNDADDENTROPY, &random_injector) == -1) {
    perror("ioctl");
  } else {
    printf("Writing %d bits of entropy to file\n", depositsize * 8);
    FILE *f;
    f = fopen("rsound.out", "a+");
    fwrite(random_injector.value.bitfield, 1, random_injector.size, f);
    fclose(f);
  }
  if (verbose > 3)
    printf("Kernel now at %d bits of entropy\n", bits_in_pool());
  if (bits_before == bits_in_pool())
    printf("Did it fail?!?!\n");
}

void
main_loop()
{
  fd_set readfds;
  fd_set writefds;
  fd_set errfds;
  int ret;
  int adding = 0;
  
  FD_ZERO(&writefds);
  
  while (1) {
    if (restart_arecord == 1) {
      int pid;
      if (verbose > 1)
        printf("Need to restart arecord.\n");
      pid = start_arecord(incoming_bits);
      if (pid == -1)
        return;
      if (verbose > 1)
        printf("Started arecord with pid %d\n", pid);
      FD_ZERO(&readfds);
      FD_ZERO(&errfds);
      if (restart_arecord) {
        printf("Arecord already died.\n");
        return;
      }
    }
    FD_SET(arecord_read_fd, &readfds);
    FD_SET(arecord_read_fd, &errfds);
    ret = select(arecord_read_fd + 1, &readfds, &writefds, &errfds, NULL);
    if (ret == -1) {
      perror("select");
      return;
    }
    if (FD_ISSET(arecord_read_fd, &readfds))
      asound_do_read();
    if (FD_ISSET(arecord_read_fd, &errfds)) {
      printf("Error on arecord fd, gotta restart.\n");
      restart_arecord = 1;
    }
    ret = transfer_bits_and_debias(incoming_bits, buffered_bits);
    if (verbose > 3 && ret > 0)
      printf("Added %d bits to cache. Now at %d/%d bits in it\n", ret, bitbuffer_available_bits(buffered_bits), buffersize * 8);
    if (bits_in_pool() <= minwatermark && adding == 0) {
      if (verbose > 2)
        printf("Transition to inserting entropy. Kernel pool at %d\n", bits_in_pool());
      adding = 1;
    }
    if (bits_in_pool() >= maxwatermark && adding == 1) {
      if (verbose > 2)
        printf("Transition to waiting. Kernel pool at %d\n", bits_in_pool());
      adding = 0;
    }
    if (adding == 1) {
      if (bitbuffer_available_bits(buffered_bits) >= (depositsize * 8)) {
        do_mixin_bits();
        if (verbose > 3)
          printf("Kernel entropy pool now sits at %d bits\n", bits_in_pool());
      }
    }
  }
}

void
usage(const char* prog, FILE *output)
{
  fprintf(output, "%s: Usage:\n"\
          "\n"\
          "Argument: h - display this help message\n"\
          "          V - display version information.\n",
          "          D - Daemonize\n"\
          "          v - Increase verbosity. Can be used more than once.\n"\
          "          m - specify minimum number of bits of entropy in the pool.\n"\
          "          M - specify max number of bits in the pool.\n"\
          "          b - specify number of bytes of randomness to buffer for use.\n"\
          "          d - specify number of bytes to deposit into the pool each time.\n",
          prog);
}

void
version(const char* prog, FILE *output)
{
  fprintf(output, "%s: %s\n"\
                  "Copyright 2007 Daniel Silverstone\n",
          prog, version_number);
}

int
main(int argc, char **argv)
{
  int opt;

  while ((opt = getopt(argc, argv, ":hDvVm:M:b:d:")) != -1) {
    switch (opt) {
    case 'h':
      usage(argv[0], stdout);
      return 0;
    case 'V':
      version(argv[0], stdout);
      return 0;
    case 'D':
      daemonise = 1;
      break;
    case 'v':
      verbose += 1;
      break;
    case 'm':
      minwatermark = atoi(optarg);
      break;
    case 'M':
      maxwatermark = atoi(optarg);
      break;
    case 'b':
      buffersize = atoi(optarg);
      break;
    case 'd':
      depositsize = atoi(optarg);
      break;
    default:
      usage(argv[0], stderr);
      return 1;
    }
  }
  
  /* Validate options now */
  if (minwatermark < 64) {
    fprintf(stderr, "Minimum watermark is below 64 bits. This is silly.\n");
    return 2;
  }
  if (maxwatermark > 4096) {
    fprintf(stderr, "Maxmimum watermark is above 4096. This is not possible.\n");
    return 2;
  }
  if (buffersize < depositsize || buffersize > (1024*1024)) {
    fprintf(stderr, "Buffer size smaller than deposit size or greater than one megabyte.\n");
    return 2;
  }
  if (depositsize < 4 || depositsize > 512 || (depositsize & 3) != 0) {
    fprintf(stderr, "Deposit size must be a multiple of four, between 4 and 512 inclusive.\n");
    return 2;
  }
  
  if (verbose > 0) {
    printf("Random sound daemon. Copyright 2007 Daniel Silverstone.\n\n");
    printf("Will keep random pool between %d and %d bits of entropy.\n", minwatermark, maxwatermark);
    printf("Will retain a buffer of %d bytes, making entropy deposits of %d bytes at a time.\n", buffersize, depositsize);
    if (daemonise == 1) {
      printf("Will daemonise.\n");
    }
  }
  
  if (verbose > 1) {
    printf("Allocating a %d bit buffer for the incoming bits.\n",
           buffersize * 8 * 4);
  }
  incoming_bits = bitbuffer_new(buffersize * 8 * 4);
  buffered_bits = bitbuffer_new(buffersize * 8);
  if (incoming_bits == NULL || buffered_bits == NULL) {
    fprintf(stderr, "Unable to allocate buffers.\n");
    return 3;
  }
  
  randomfd = open("/dev/random", O_RDWR);
  
  if (randomfd == -1) {
    perror("Opening /dev/random\n");
    return 3;
  }
  
  if (daemonise == 1) {
    printf("Daemonising\n");
    fflush(stdout);
    int fd = fork();
    if (fd == -1 ) {
      perror("forking daemon.\n");
      return 4;
    }
    if (fd != 0) return 0;
    setpgrp();
    setsid();
  }
  
  main_loop();
  
  return -1;
}
